File: backend/app/modules/auth/service.py
"""Authentication and user provisioning."""
from __future__ import annotations

from app.core.errors import AuthenticationError, ConflictError
from app.core.permissions import Role
from app.core.security import (
    create_access_token,
    create_refresh_token,
    decode_token,
    hash_password,
    verify_password,
)
from app.db.models import Organization, User
from app.db.repositories import UserRepository
from app.db.session import UnitOfWork


class AuthService:
    def __init__(self, session: UnitOfWork) -> None:
        self.session = session
        self.users = UserRepository(session)

    def create_organization(self, name: str) -> Organization:
        organization = Organization(name=name)
        self.session.add(organization)
        self.session.flush()
        return organization

    def create_user(
        self, organization_id: str, email: str, password: str, role: Role | str
    ) -> User:
        email = email.lower()
        if self.users.by_email(email) is not None:
            raise ConflictError("A user with that email already exists.", email=email)
        user = User(
            organization_id=organization_id,
            email=email,
            hashed_password=hash_password(password),
            role=str(Role(role)),
        )
        self.users.add(user)
        return user

    def login(self, email: str, password: str) -> tuple[str, str]:
        user = self.users.by_email(email.lower())
        if user is None or not verify_password(password, user.hashed_password):
            # Same error either way: never reveal whether an account exists.
            raise AuthenticationError("Invalid email or password.")
        return (
            create_access_token(user.id, user.organization_id, user.role),
            create_refresh_token(user.id, user.organization_id, user.role),
        )

    def refresh(self, refresh_token: str) -> tuple[str, str]:
        payload = decode_token(refresh_token, expected_type="refresh")
        return (
            create_access_token(str(payload["sub"]), str(payload["org"]), str(payload["role"])),
            create_refresh_token(str(payload["sub"]), str(payload["org"]), str(payload["role"])),
        )

File: apps/web/lib/server/auth/service.ts
/**
 * Authentication and user provisioning. Port of backend/app/modules/auth/service.py.
 */
import { AuthenticationError, ConflictError } from "@/lib/server/errors";
import { UserRepository } from "@/lib/server/db/repositories";
import type { UnitOfWork } from "@/lib/server/db/session";
import { newOrgOwned, newEntity } from "@/lib/server/db/base";
import type { Organization, User } from "@/lib/server/db/models";
import {
  createAccessToken,
  createRefreshToken,
  decodeToken,
  hashPassword,
  verifyPassword,
} from "./security";

export interface TokenPair {
  access_token: string;
  refresh_token: string;
  token_type: "bearer";
}

async function issueTokens(userId: string, orgId: string): Promise<TokenPair> {
  const [access_token, refresh_token] = await Promise.all([
    createAccessToken(userId, orgId),
    createRefreshToken(userId, orgId),
  ]);
  return { access_token, refresh_token, token_type: "bearer" };
}

export class AuthService {
  private users: UserRepository;

  constructor(private session: UnitOfWork) {
    this.users = new UserRepository(session);
  }

  async createOrganization(name: string): Promise<Organization> {
    const organization: Organization = { ...newEntity(), name };
    this.session.add("organizations", organization);
    await this.session.flush();
    return organization;
  }

  async createUser(organizationId: string, email: string, password: string): Promise<User> {
    const normalizedEmail = email.toLowerCase();
    if ((await this.users.byEmail(normalizedEmail)) !== null) {
      throw new ConflictError("A user with that email already exists.", { email: normalizedEmail });
    }
    const user: User = {
      ...newOrgOwned(organizationId),
      email: normalizedEmail,
      hashedPassword: await hashPassword(password),
      isActive: true,
    };
    this.users.add(user);
    return user;
  }

  async login(email: string, password: string): Promise<TokenPair> {
    const user = await this.users.byEmail(email.toLowerCase());
    // Same error either way: never reveal whether an account exists.
    if (!user || !(await verifyPassword(password, user.hashedPassword))) {
      throw new AuthenticationError("Invalid email or password.");
    }
    return issueTokens(user.id, user.organizationId);
  }

  async refresh(refreshToken: string): Promise<TokenPair> {
    const payload = await decodeToken(refreshToken, "refresh");
    return issueTokens(payload.sub, payload.org);
  }
}
